File: client-examples/csharp/NullAuthClient.cs
using System;
using System.Diagnostics;
using System.Net.Http;
using System.Security.Cryptography;
using System.Security.Principal;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace NullAuthClient
{
    public class NullAuthResult
    {
        public bool Success { get; set; }
        public string Message { get; set; }
        public string ErrorCode { get; set; }
        public NullAuthData Data { get; set; }
    }

    public class NullAuthData
    {
        public string Status { get; set; }
        public string ExpiresAt { get; set; }
        public int RemainingDays { get; set; }
        public string FirstActivatedAt { get; set; }
    }

    public class NullAuthSDK
    {
        private readonly string _baseUrl;
        private readonly string _appId;
        private readonly string _appSecret;
        private static readonly HttpClient _httpClient = new HttpClient();

        public NullAuthSDK(string appId, string appSecret, string baseUrl = "https://null-auth-backend.vercel.app")
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _appId = appId;
            _appSecret = appSecret;
        }

        /// <summary>
        /// Retrieves the Windows User SID using legitimate 'whoami /user' fallback to WindowsIdentity.
        /// </summary>
        public static string GetWindowsUserSid()
        {
            try
            {
                ProcessStartInfo psi = new ProcessStartInfo("whoami", "/user")
                {
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };

                using (Process process = Process.Start(psi))
                {
                    string output = process.StandardOutput.ReadToEnd();
                    process.WaitForExit();

                    string[] lines = output.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
                    foreach (string line in lines)
                    {
                        if (line.Contains("S-1-5-"))
                        {
                            string[] parts = line.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                            if (parts.Length >= 2)
                            {
                                return parts[parts.Length - 1].Trim();
                            }
                        }
                    }
                }
            }
            catch { }

            try
            {
                return WindowsIdentity.GetCurrent().User?.Value ?? "UNKNOWN_SID";
            }
            catch
            {
                return "UNKNOWN_HWID";
            }
        }

        /// <summary>
        /// Authenticates using License Key (App Mode 1)
        /// </summary>
        public async Task<NullAuthResult> AuthenticateLicenseAsync(string licenseKey)
        {
            string hwid = GetWindowsUserSid();
            string endpoint = $"{_baseUrl}/api/v1/client/license/authenticate";

            var payload = new
            {
                appId = _appId,
                appSecret = _appSecret,
                licenseKey = licenseKey?.Trim(),
                hwid = hwid
            };

            return await SendAuthRequestAsync(endpoint, payload);
        }

        /// <summary>
        /// Authenticates using direct HWID Whitelist Access (App Mode 2)
        /// </summary>
        public async Task<NullAuthResult> AuthenticateHwidAsync()
        {
            string hwid = GetWindowsUserSid();
            string endpoint = $"{_baseUrl}/api/v1/client/hwid/authenticate";

            var payload = new
            {
                appId = _appId,
                appSecret = _appSecret,
                hwid = hwid
            };

            return await SendAuthRequestAsync(endpoint, payload);
        }

        private async Task<NullAuthResult> SendAuthRequestAsync(string endpoint, object payload)
        {
            try
            {
                string jsonBody = JsonSerializer.Serialize(payload);
                var content = new StringContent(jsonBody, Encoding.UTF8, "application/json");

                HttpResponseMessage response = await _httpClient.PostAsync(endpoint, content);
                string responseString = await response.Content.ReadAsStringAsync();

                using JsonDocument doc = JsonDocument.Parse(responseString);
                JsonElement root = doc.RootElement;

                bool success = root.GetProperty("success").GetBoolean();
                string message = root.GetProperty("message").GetString();
                string errorCode = root.TryGetProperty("error", out var errElem) ? errElem.GetString() : null;

                NullAuthData data = null;
                if (success && root.TryGetProperty("data", out var dataElem))
                {
                    data = new NullAuthData
                    {
                        Status = dataElem.TryGetProperty("status", out var s) ? s.GetString() : "active",
                        ExpiresAt = dataElem.TryGetProperty("expires_at", out var e) ? e.GetString() : "",
                        RemainingDays = dataElem.TryGetProperty("remaining_days", out var r) ? r.GetInt32() : 0,
                        FirstActivatedAt = dataElem.TryGetProperty("first_activated_at", out var f) ? f.GetString() : null,
                    };
                }

                return new NullAuthResult
                {
                    Success = success,
                    Message = message,
                    ErrorCode = errorCode,
                    Data = data
                };
            }
            catch (Exception ex)
            {
                return new NullAuthResult
                {
                    Success = false,
                    Message = "Connection failed to Null-Auth server.",
                    ErrorCode = "NETWORK_ERROR",
                    Data = null
                };
            }
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            Console.Title = "Null-Auth C# Integration Client";
            Console.ForegroundColor = ConsoleColor.Red;
            Console.WriteLine("=================================================");
            Console.WriteLine("        Null-Auth C# Integration Sample          ");
            Console.WriteLine("=================================================");
            Console.ResetColor();

            // Set your App ID and Secret Key from Null-Auth Dashboard
            string appId = "NA-48392017";
            string appSecret = "nas_YOUR_APP_SECRET_HERE";
            string apiUrl = "https://null-auth-backend.vercel.app";

            string userSid = NullAuthSDK.GetWindowsUserSid();
            Console.WriteLine($"[+] Detected Windows User SID: {userSid}");

            var sdk = new NullAuthSDK(appId, appSecret, apiUrl);

            Console.Write("\nEnter License Key (e.g. NULL-ABCD-1234-EFGH): ");
            string licenseKey = Console.ReadLine();

            Console.WriteLine("\n[*] Authenticating with Null-Auth Cloud Server...");
            NullAuthResult result = await sdk.AuthenticateLicenseAsync(licenseKey);

            if (result.Success)
            {
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("\n[+] ACCESS GRANTED! Application Unlocked.");
                Console.WriteLine($"    Status: {result.Data.Status}");
                Console.WriteLine($"    Expires At: {result.Data.ExpiresAt}");
                Console.WriteLine($"    Remaining Days: {result.Data.RemainingDays} days");
                Console.ResetColor();
            }
            else
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("\n[-] ACCESS DENIED!");
                Console.WriteLine($"    Reason: {result.Message}");
                Console.WriteLine($"    Error Code: {result.ErrorCode}");
                Console.ResetColor();
            }

            Console.WriteLine("\nPress any key to exit...");
            Console.ReadKey();
        }
    }
}

File: client-examples/python/null_auth_client.py
#!/usr/bin/env python3
"""
Null-Auth Python Client SDK
Zero-dependency Python 3 integration client for Null-Auth Private Licensing & Auth Platform.
"""

import sys
import subprocess
import json
import urllib.request
import urllib.error

class NullAuthSDK:
    def __init__(self, app_id: str, app_secret: str, api_base_url: str = "https://null-auth-backend.vercel.app"):
        self.api_base_url = api_base_url.rstrip('/')
        self.app_id = app_id
        self.app_secret = app_secret

    @staticmethod
    def get_windows_user_sid() -> str:
        """
        Retrieves the current Windows User SID using legitimate 'whoami /user' command.
        """
        try:
            output = subprocess.check_output(["whoami", "/user"], text=True)
            for line in output.splitlines():
                if "S-1-5-" in line:
                    parts = line.split()
                    if parts:
                        return parts[-1].strip()
        except Exception:
            pass

        return "UNKNOWN_HWID"

    def authenticate_license(self, license_key: str) -> dict:
        """
        Authenticate client using App ID + Secret + License Key + Windows HWID
        """
        hwid = self.get_windows_user_sid()
        url = f"{self.api_base_url}/api/v1/client/license/authenticate"
        
        payload = {
            "appId": self.app_id,
            "appSecret": self.app_secret,
            "licenseKey": license_key.strip(),
            "hwid": hwid
        }

        return self._send_request(url, payload)

    def authenticate_hwid(self) -> dict:
        """
        Authenticate client using App ID + Secret + Windows HWID (HWID Whitelist Mode)
        """
        hwid = self.get_windows_user_sid()
        url = f"{self.api_base_url}/api/v1/client/hwid/authenticate"
        
        payload = {
            "appId": self.app_id,
            "appSecret": self.app_secret,
            "hwid": hwid
        }

        return self._send_request(url, payload)

    def _send_request(self, url: str, payload: dict) -> dict:
        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json", "User-Agent": "NullAuthPythonSDK/1.0"},
            method="POST"
        )

        try:
            with urllib.request.urlopen(req) as response:
                res_data = response.read().decode('utf-8')
                return json.loads(res_data)
        except urllib.error.HTTPError as e:
            error_data = e.read().decode('utf-8')
            try:
                return json.loads(error_data)
            except Exception:
                return {"success": False, "message": f"HTTP Error {e.code}", "error": f"HTTP_{e.code}"}
        except Exception as e:
            return {"success": False, "message": "Failed to connect to Null-Auth cloud server.", "error": "NETWORK_ERROR"}

if __name__ == "__main__":
    print("=================================================")
    print("       Null-Auth Python Integration SDK          ")
    print("=================================================")

    # Configure your App ID and Secret Key from Null-Auth Dashboard
    API_URL = "https://null-auth-backend.vercel.app"
    APP_ID = "NA-13026130"
    APP_SECRET = "nas_a11d5a6604bec0b8227a8fbbc472a10722b27b66014a2634"

    sdk = NullAuthSDK(APP_ID, APP_SECRET, API_URL)

    sid = sdk.get_windows_user_sid()
    print(f"[+] Detected Windows User SID: {sid}")

    license_key = input("\nEnter License Key (e.g. NULL-ABCD-1234-EFGH): ").strip()

    print("\n[*] Sending authentication request...")
    result = sdk.authenticate_license(license_key)

    print("\n[Result]:")
    print(json.dumps(result, indent=2))

    if result.get("success"):
        print(f"\n[+] ACCESS GRANTED! Status: {result.get('data', {}).get('status')}")
    else:
        print(f"\n[-] ACCESS DENIED! Reason: {result.get('message')}")
